File: Advent2016/Day/Day_08.py
fd = open("""../Resources/Day_08.txt""", 'r')
s = fd.read()
fd.close()

words = s.split('\n')

# # - 1 - on 
# . - 0 - off
#50x6
screen = [ [ 0 for i in range(6)] for i in range(50)]

#instructions
#rect - switches on the lights
#rotate col (Y coord)
#rotate row (X coord)

ex = 0
for i in words:
    rot = 0;
    
    if (i.find("row") != -1):
        s = i.split(' ')
        by = int(s[-1])
        y = int(s[2].split('=')[1])
        t = list(map(lambda z: z[y],screen))
        t = t[-by:] + t[:-by]
        for j in range(50):
            screen[j][y] = t[j]
        
        
    if (i.find("column") != -1):
        s = i.split(' ')
        by = int(s[-1])
        x = int(s[2].split('=')[1])
        screen[x] = screen[x][-by:] + screen[x][:-by]

    #XxY
    if (i.find("rect") != -1):
        xy = i.split(' ')[1].split('x')
        x = int(xy[0])
        y = int(xy[1])
        for I in range(x):
            for J in range(y):
                screen[I][J] = 1

#task 1
print(sum([sum(i) for i in screen]))

#task 2
a = ''
for i in range(6):
	for j in range(50):
		a+= screen[j][i] and '#' or ' '
	print(a)
	a = ''

    

File: Advent2016/Day/Day_08.cpp
#include "stdafx.h"
#include "Day_08.h"

Day_08::Day_08::Day_08() 
: task1(0)
{
	for (size_t x = 0; x < WIDTH; ++x) {
		for (size_t y = 0; y < HEIGHT; ++y) {
			screen[x][y] = '.'; // off
		}
	}
};

Day_08::Day_08::~Day_08() {
};

unsigned short Day_08::Day_08::Count() {
	unsigned short cnt = 0;
	for (size_t x = 0; x < WIDTH; ++x) {
		for (size_t y = 0; y < HEIGHT; ++y) {
			cnt += (screen[x][y] == '#') ? 1 : 0; 
		}
	}
	return cnt;
}

void Day_08::Day_08::ExecuteInstruction(std::vector<std::string> instr) {
	if (instr.size() < 1)
		return;
	if (instr[0] == "rect") {
		size_t xPos = instr[1].find('x');
		size_t x = atoi(instr[1].substr(0, xPos).c_str());
		size_t y = atoi(instr[1].substr(xPos+1).c_str());
		SwitchOn(x, y);

	} else if (instr[1] == "column") {
		size_t pos = instr[2].find('=');
		size_t x = atoi(instr[2].substr(pos+1).c_str());
		RotateColumn(x, atoi(instr[4].c_str()));

	} else if (instr[1] == "row") {
		size_t pos = instr[2].find('=');
		size_t y = atoi(instr[2].substr(pos+1).c_str());
		RotateRow(y, atoi(instr[4].c_str()));
	}
}

void Day_08::Day_08::SwitchOn(size_t X, size_t Y) {
	for (size_t x = 0; x < X; ++x) {
		for (size_t y = 0; y < Y; ++y) {
			screen[x][y] = '#';
		}
	}
}

void Day_08::Day_08::RotateColumn(size_t X, size_t by) {
	char tmp[HEIGHT];

	for (int y = 0; y < HEIGHT; ++y) {
		tmp[y] = screen[X][y];
	}

	for (int y = 0; y < HEIGHT; ++y) {
		screen[X][(y + by)%HEIGHT] = tmp[y];
	}
};

void Day_08::Day_08::RotateRow(size_t Y, size_t by) {
	char tmp[WIDTH];

	for (int x = 0; x < WIDTH; ++x) {
		tmp[x] = screen[x][Y];
	}

	for (int x = 0; x < WIDTH; ++x) {
		screen[(x + by)%WIDTH][Y] = tmp[x];
	}
};

void Day_08::Day_08::Evaluate() {
	std::string wordPart;
	std::vector<std::string> instruction;
	
	size_t idx = 0;

	for (auto& word : words) {
		if (word.compare("rect") == 0
			|| word.compare("rotate") == 0) {
			//new instruction execute last one
			ExecuteInstruction(instruction);
			instruction.clear();

		} 
		instruction.push_back(word);
	}
}

void Day_08::Day_08::Print() {
	std::cout << "Day 08 task1: " << Count() << std::endl;
	for (int y = 0; y < HEIGHT; ++y) {
		for (int x = 0; x < WIDTH; ++x) {
			std::cout << (screen[x][y] == '#' ? '#' : ' ');
		}
		std::cout << std::endl;
	}
}